lib/spack/spack/build_systems/go.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+from llnl.util.filesystem import mkdirp, install
+from spack.directives import depends_on
+from spack.package import PackageBase, run_after
+from spack.util.executable import Executable
+
+
+class GoPackage(PackageBase):
+    """Specialized class for packages that are built using Go.
+
+    This class provides two phases that can be overridden if required:
+
+        1. :py:meth:`~.GoPackage.build`
+        2. :py:meth:`~.GoPackage.install`
+
+    Packages must add the binaries to installed to
+    :py:attr:`~.GoPackage.executables`.
+
+    Additional build flags can be added to
+    :py:attr:`~.GoPackage.build_flags`.
+
+    :py:meth:`~.GoPackage.setup_build_environment` sets
+    ``GO111MODULE=on`` to enable module mode and sets
+    ``GOFLAGS="-mod-vendor"`` to enable vendoring and prevent ``go``
+    from downloading libraries at build time.
+    """
+
+    #: Phases of a Go package
+    phases = ['build', 'install']
+
+    #: This attribute is used in UI queries that need to know the build
+    #: system base class
+    build_system_class = 'GoPackage'
+
+    depends_on('go@1.11:', type=('build'))
```

Comment:
All projects require 1.11+?

Reply:
Yes.  go@1.11 is the version where modules support was introduced, though many features were in flux.  The current release (@1.13) is the release where they think the interface is stable and production ready.  For our purposes (`-mod=vendor`) the support in `@1.11` is sufficient.  Sounds like the old project structure (aka GOPATH) will be deprecated in either 1.14 or 1.15 and eventually removed.

Go only supports the previous two releases, so even `@1.11` will be out of support fairly soon.

[I think that] it's not a problem, given that it's so easy to build a current toolchain (a couple of minutes on my mac laptop), that Go's backward compatibility is so good, and that it's only a *build* time dependency with no-ish run time implications.

If there is an odd-ball that insists on an older Go compiler, there's nothings stopping it.  But
since it can't take advantage of the way `GoPackage` forces `go build` to use the vendor dir, it's going to need to redo most of it anyway, so....